### PR TITLE
Various inferrability improvements

### PR DESCRIPTION
**Pull request by timholy**
- two functions affected by `Core.Box` (#15276) were ameliorated
- type info added to the `stack` field of `ExceptionStack`
- avoid unknown dispatch from de-specialized code

### CODE REVIEW

File: stdlib/Test/src/Test.jl [hunk 1]
```diff
@@ -1580,7 +1580,10 @@ function _inferred(ex, mod, allow = :(Union{}))
         ex = Expr(:call, GlobalRef(Test, :_materialize_broadcasted),
             farg, ex.args[2:end]...)
     end
-    Base.remove_linenums!(quote
+    Base.remove_linenums!(_inferred_gen(ex))
+end
+function _inferred_gen(ex)   # split from the above for julia#15276
+    quote
```
[the remark this answers is not in the document]
Comment (vtjnash) [reply stdlib/Test/src/Test.jl]: you'll need to fix up the end too, for the extra `end` and `)`

File: stdlib/REPL/src/docview.jl [hunk 2]
```diff
@@ -402,8 +402,9 @@ function symbol_latex(s::String)
 
     return get(symbols_latex, s, "")
 end
-function repl_latex(io::IO, s::String)
-    latex = symbol_latex(s)
+function repl_latex(io::IO, s0::String)
+    s::String = s0
```
Comment (rfourquet) [inline stdlib/REPL/src/docview.jl:406]: This seems really mysterious to me, maybe add a comment? (sorry if this is obvious, I'm not up-to-date with this kind of improvements :) )

File: base/arrayshow.jl [hunk 3]
```diff
@@ -95,7 +96,8 @@ is specified as string sep.
 """
 function print_matrix_row(io::IO,
         @nospecialize(X::AbstractVecOrMat), A::Vector,
-        i::Integer, cols::AbstractVector, sep::AbstractString)
+        i::Integer, cols::AbstractVector, sep::AbstractString,
+        idxlast::Integer=last(axes(X, 2)))
```
Comment (timholy) [inline base/arrayshow.jl:100]:
```suggestion
        #= `axes(X)` may not infer, set this in caller =# idxlast::Integer=last(axes(X, 2)))
```

File: stdlib/REPL/src/docview.jl [hunk 4]
```diff
@@ -402,8 +402,9 @@ function symbol_latex(s::String)
 
     return get(symbols_latex, s, "")
 end
-function repl_latex(io::IO, s::String)
-    latex = symbol_latex(s)
+function repl_latex(io::IO, s0::String)
```
Comment (timholy) [inline stdlib/REPL/src/docview.jl:405]:
```suggestion
function repl_latex(io::IO, s0::String)
    # This has rampant `Core.Box` problems (#15276). Use the tricks of
    # https://docs.julialang.org/en/v1/manual/performance-tips/#man-performance-captured
    # We're changing some of the values so the `let` trick isn't applicable. 
```

**Review by timholy — COMMENTED** [review]
@rfourquet, better?

File: stdlib/REPL/src/docview.jl (hunk 4 above)
Comment (rfourquet) [reply stdlib/REPL/src/docview.jl:405]: Is the problem here linked to the fact that a `String` is not a real immutable object under the hood?

Comment (rfourquet) [reply stdlib/REPL/src/docview.jl:405]: >  better?

Yes that's great, thanks!

Comment (timholy) [reply stdlib/REPL/src/docview.jl:405]: No, it's just that it's captured by the closure *and* set in the main body.

**Comment by DilumAluthge** [comment]
Just FYI, this probably should not have been merged with failing CI.

For example, the `whitespace` failure is caused by trailing whitespace introduced in this PR, specifically on line 408 of the `stdlib/REPL/src/docview.jl ` file.

The trailing whitespace will be removed in #41998.

**Comment by timholy** [comment]
Agreed, sorry I didn't notice the whitespace failure, and thanks for fixing it.

**Comment by DilumAluthge** [comment]
No worries at all! Whitespace failures are the easiest CI failures to fix 😄

**Pull request merged**